gcc/common/config/riscv/riscv-common.c

```diff
@@ -68,6 +68,8 @@ riscv_implied_info_t riscv_implied_info[] =
   {"zks", "zksh"},
   {"zks", "zkg"},
   {"zks", "zkb"},
+  {"p", "zpn"},
+  {"p", "zpsf"},
```

Comment:
zprv?

Reply:
The mapping here means the `p` flag in `--with-arch=rv{xlen}gcp` or `-march=rv{xlen}gcp` can be expanded to `rv{xlen}gc_zpn_zpsf`.

Reply:
I agree. Still I think that zprv should appear on xlen=64.
When I run `riscv64-unknown-elf-gcc -march=rv64imcbp -mabi=lp64 .....` on compiler with this PR, I get
```
rv64up.c: warning: implicit declaration of function '__rv__kadd32' [-Wimplicit-function-declaration]
    |   if (0x0000000000000000 != __rv__kadd32(0x0000000000000000, 0x0000000000000000)) return 536;
    |                             ^~~~~~~~~~~~
```

This is so because `rv64imcbp` expands to `rv64imcb_zpn_zpsf` whilest it should expand to `rv64imcb_zpn_zpsf_zprv` because `zprv` is required on `xlen=64` by spec (I simplify a bit: I didn't expand b to its zeds here because it is not the issue).
If I do not treat warnings as errors, it obviously fails on linking with:
```
objs/rv64imcb@pcsh-bm1_rv64up/rv64up.o: in function `.L0 ': undefined reference to `__rv__kadd32'
```
I faced a more general problem with this PR: when I compile my code with `riscv64-unknown-elf-gcc -march=rv64imcb -mabi=lp64 .....` (i.e. **without** p-ext) it fails on linking with this error:
```
objs/rv64imcb@pcsh-bm1_rv64up/rv64up.o: in function `.L0 ': undefined reference to `__builtin_riscv_kadd16'
```

I have my own p-ext intrinsic file (which is generally
```cpp
#ifdef __riscv_zpn
static inline __rv__insn(uintXLEN_t rs1, uintXLEN_t rs2) { uintXLEN_t rd; __asm__("insn %0, %1, %2" : "=r"(rd) : "r"(rs1), "r"(rs2)); return rd; }
#else
static inline __rv__insn(uintXLEN_t rs1, uintXLEN_t rs2) { SOME_GENERIC_CODE; return rd; }
#endif
```
with different variations of types, number of arguments, clobbers and other stuff). I did so because I am not a pro at machine description language and also it seems to be a common way to implement intrinsics (e.g. https://github.com/rvkrypto/rvkrypto-fips/blob/main/rvkintrin.h).

So my intrinsics produce an assembler code with the specified insn if p-ext is enabled and make a call to (or inline -- depends on optimization levels) some generic-code functions otherwise.

When I found this PR, I expected it works the same way and in addition allows to optimize some generic code with emitting of p-ext insns.
As you mentioned in another topic, "generic code" should use very specific types defined in the same file where intrinsics are -- so we barely expect that such insns will appear on some code with common scalar types as b-ext insns do. It is ok for me: I understand that b-ext machine descriptions implement some simple stuff (e.g. bit set) and this PR operates with packed types so this is either very hard or impossible to map a set of scalar type insns to p-ext insn in most cases. Still I don't understand why intrinsics do not work (I mean do not provide generic code) without enabled p-ext and what should be done if I want to compile some file with the compiler which includes this intrinsic file without p-ext insns.

Reply:
Hi @marcfedorow,

> This is so because rv64imcbp expands to rv64imcb_zpn_zpsf whilest it should expand to rv64imcb_zpn_zpsf_zprv because zprv is required on xlen=64 by spec (I simplify a bit: I didn't expand b to its zeds here because it is not the issue).

Thank you for pointing it out. I add a fix to support a different expansion on p flag (rv32p will be expanded into rv32_zpn_zpsf, and rv64p will be rv64_zpn_zprv_zpsf).

> with different variations of types, number of arguments, clobbers and other stuff). I did so because I am not a pro at machine description language and also it seems to be a common way to implement intrinsics (e.g. https://github.com/rvkrypto/rvkrypto-fips/blob/main/rvkintrin.h).
So my intrinsics produce an assembler code with the specified insn if p-ext is enabled and make a call to (or inline -- depends on optimization levels) some generic-code functions otherwise.

I think we do not have such a plan so far, since it requires a great amount of workload. I think it is a good idea though, so I will forward your message to other PLCT members in the further meetup. 

> As you mentioned in another topic, "generic code" should use very specific types defined in the same file where intrinsics are -- so we barely expect that such insns will appear on some code with common scalar types as b-ext insns do. It is ok for me: I understand that b-ext machine descriptions implement some simple stuff (e.g. bit set) and this PR operates with packed types so this is either very hard or impossible to map a set of scalar type insns to p-ext insn in most cases. Still I don't understand why intrinsics do not work (I mean do not provide generic code) without enabled p-ext and what should be done if I want to compile some file with the compiler which includes this intrinsic file without p-ext insns.

Vector type `int16x4_t`  doesn't need to be from p-ext header, since it is native support in GCC (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html). To use intrinsic might be the only solution if you try to use scalar type instead of vector type (e.g. put int16x4_t data onto int64_t field), but p-ext insn is possible to be generated from scalar code through auto-vectorization. e.g.

```c
#include <rvp_intrinsic.h>
#include <stdint.h>

typedef short v4hi __attribute__((vector_size (8)));

v4hi v_sadd16_spn (v4hi ra, v4hi rb)
{
  return ra + rb;
}

int16_t *v_sadd16_arr (int16_t *ra, int16_t *rb, int len)
{
  for (int i = 0; i < len; i++)
	  ra[i] += rb[i];
  return ra;
}
```

riscv64-unknown-elf-gcc -S -O3 add16.c
```
	.file	"add16.c"
	.option nopic
	.attribute arch, "rv64i2p0_m2p0_a2p0_f2p0_d2p0_c0p0_p2p0_zpn2p0_zprv2p0_zpsf2p0"
	.attribute unaligned_access, 0
	.attribute stack_align, 16
	.text
	.align	1
	.globl	v_sadd16_spn
	.type	v_sadd16_spn, @function
v_sadd16_spn:
	add16	a0, a0, a1
	ret
	.size	v_sadd16_spn, .-v_sadd16_spn
	.align	1
	.globl	v_sadd16_arr
	.type	v_sadd16_arr, @function
v_sadd16_arr:
	ble	a2,zero,.L4
	addi	a5,a1,2
	addiw	a3,a2,-1
	sub	a5,a0,a5
	sext.w	a4,a3
	sltiu	a5,a5,5
	li	a6,5
	xori	a5,a5,1
	sgtu	a4,a4,a6
	and	a5,a4,a5
	sext.w	t1,a2
	beq	a5,zero,.L5
	or	a5,a1,a0
	andi	a5,a5,7
	bne	a5,zero,.L5
	srliw	a7,t1,2
	slli	a7,a7,3
	mv	a5,a0
	mv	a3,a1
	add	a7,a7,a0
.L6:
	ld	a4,0(a5)
	ld	a6,0(a3)
	addi	a5,a5,8
	addi	a3,a3,8
	add16	a4, a4, a6
	sd	a4,-8(a5)
	bne	a5,a7,.L6
	andi	a5,t1,-4
	mv	a4,a5
	beq	t1,a5,.L4
	slli	a5,a5,32
	srli	a5,a5,31
	add	a7,a0,a5
	add	a3,a1,a5
	lhu	a6,0(a3)
	lhu	t1,0(a7)
	addiw	a3,a4,1
	addw	a6,a6,t1
	sh	a6,0(a7)
	bge	a3,a2,.L4
	addi	a3,a5,2
	add	a6,a0,a3
	add	a3,a1,a3
	lhu	a3,0(a3)
	lhu	a7,0(a6)
	addiw	a4,a4,2
	addw	a3,a3,a7
	sh	a3,0(a6)
	bge	a4,a2,.L4
	addi	a5,a5,4
	add	a4,a0,a5
	add	a1,a1,a5
	lhu	a5,0(a1)
	lhu	a3,0(a4)
	addw	a5,a5,a3
	sh	a5,0(a4)
	ret
.L5:
	slli	a5,a3,32
	srli	a3,a5,31
	addi	a2,a0,2
	mv	a5,a0
	add	a2,a2,a3
.L8:
	lhu	a4,0(a5)
	lhu	a3,0(a1)
	addi	a5,a5,2
	addi	a1,a1,2
	addw	a4,a4,a3
	sh	a4,-2(a5)
	bne	a5,a2,.L8
.L4:
	ret
	.size	v_sadd16_arr, .-v_sadd16_arr
	.ident	"GCC: (GNU) 10.2.0"
```

I hope I have answered all your questions. Thanks again.